media_driver/linux/common/ddi/media_ddi_render_target_table.cpp

```diff
@@ -0,0 +1,362 @@
+//!
+
+#include "media_ddi_render_target_table.h"
+#include "media_libva.h"
+#include "media_libva_util.h"
+
+#include <algorithm>
+#include <vector>
+
+//!
+//! \brief    Init
+//! \details  Initialize the render target table. Should be called prior to other interactions
+//!           with the render target table. Since the input parameter is codec-specific,
+//!           Init should most likely be called in the codec-specific media context initialization
+//!           stage.
+//!
+//! \param    [in] max_num_entries
+//!           Maximum number of render targets to track in the render target table. Should realistically
+//!           not exceed the maximum number of the uncompressed surface buffer of the driver.
+//!
+
+void DDI_CODEC_RENDER_TARGET_TABLE::Init(size_t max_num_entries)
+{
+    m_current_rt_surface = VA_INVALID_ID;
+    m_current_recon_target = VA_INVALID_ID;
+    m_free_index_pool.clear();
+    m_va_to_rt_map.clear();
+
+    for (RTTableIdx i = 0; i < max_num_entries; i++)
+    {
+        m_free_index_pool.push_back(i);
+    }
+}
+//!
+//! \brief    Register Render Target Surface
+//! \details  Register surface in render target table
+//!
+//! \param    [in] id
+//!           VASurfaceID for the render target to be registered. If there is no more place
+//!           inside the render target table, one inactive render target will be evicted in order
+//!           to put the new render target in its place, and an error status returned.
+//!
+//! \return   VAStatus
+//!           VA_STATUS_SUCCESS if successful (either id has been registered, or had
+//!           already been registered), VA_STATUS_ERROR_MAX_NUM_EXCEEDED if a surface eviction was
+//!           required to register the new render target, VA_STATUS_ERROR_NOT_ENOUGH_BUFFER if it is
+//!           impossible to register the new render target because none are marked as inactive.
+//!
+//!
+
+VAStatus DDI_CODEC_RENDER_TARGET_TABLE::RegisterRTSurface(VASurfaceID id)
+{
+    if (id == VA_INVALID_ID)
+    {
+        DDI_ASSERTMESSAGE("Invalid VASurfaceID in RegisterRTSurfaces");
+        return VA_STATUS_ERROR_INVALID_PARAMETER;
+    }
+
+    if (IsRegistered(id))
+    {
+        return VA_STATUS_SUCCESS;
+    }
+
+    if (!m_free_index_pool.empty())
+    {
+        RTTableIdx idx = m_free_index_pool.front();
+        m_free_index_pool.pop_front();
+
+        DDI_CODEC_RENDER_TARGET_INFO info;
+        info.FrameIdx = idx;
+        info.RTState = RT_STATE_ACTIVE_IN_CURFRAME;
+
+        m_va_to_rt_map[id] = info;
+        return VA_STATUS_SUCCESS;
+    }
+    else
+    {
+        using RTPair = std::pair<VASurfaceID, DDI_CODEC_RENDER_TARGET_INFO>;
+        using RTPairIterator = std::map<VASurfaceID, DDI_CODEC_RENDER_TARGET_INFO>::iterator;
+        std::vector<RTPairIterator> inactive_rt_vec;
+        inactive_rt_vec.reserve(m_va_to_rt_map.size());
+
+        for (RTPairIterator it = m_va_to_rt_map.begin(); it != m_va_to_rt_map.end(); it++)
```

Comment:
We can find the smallest `FrameIdx` right in this loop, why we need additional `inactive_rt_vec` + `min_element`?

Reply:
Indeed, will re-write this section.